### PR TITLE
Letter templates

### PR DESCRIPTION
**Pull request by elena-vi**

### CODE REVIEW

File: lib/hackney/pdf/generate_preview.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+module Hackney
+  module PDF
+    class GeneratePreview
```
Comment (elena-vi) [inline lib/hackney/pdf/generate_preview.rb]: preview not generate preview

File: lib/hackney/pdf/generate_preview.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      def initialize(get_templates_gateway:, get_case_by_refs_gateway:)
+        @get_templates_gateway = get_templates_gateway
+        @get_case_by_refs_gateway = get_case_by_refs_gateway
+      end
+
+      def execute(payment_ref:, template_id:)
+        template = get_template_by_id(template_id)
+        sc_case = @get_case_by_refs_gateway.execute(payment_ref: payment_ref).first
```
Comment (elena-vi) [inline lib/hackney/pdf/generate_preview.rb]: get case

Comment (elena-vi) [reply lib/hackney/pdf/generate_preview.rb]: move to method

File: lib/hackney/pdf/generate_preview.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+module Hackney
+  module PDF
+    class GeneratePreview
+      def initialize(get_templates_gateway:, get_case_by_refs_gateway:)
```
Comment (elena-vi) [inline lib/hackney/pdf/generate_preview.rb]: `get_case_by_refs_gateway` no more case, use `leasehold_info`

File: lib/hackney/pdf/get_templates.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+      def get_meta_data(file_path)
+        {
+          name: get_template_name(file_path),
+          id: get_template_id(file_path)
+        }
+      end
+
+      def get_template_name(file_path)
+        file_path.split('/').last.split('.').first.tr('_', ' ').capitalize
```
Comment (alexdesi) [inline lib/hackney/pdf/get_templates.rb]: this is a bit complicated,
you coud use https://ruby-doc.org/core-2.1.5/File.html#method-c-basename
https://ruby-doc.org/core-2.1.5/File.html#method-c-extname
and other similar method

File: lib/hackney/pdf/get_templates.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      end
+
+      def get_template_id(file_path)
+        file_path.split('/').last.split('.').first
```
Comment (alexdesi) [inline lib/hackney/pdf/get_templates.rb]: as before.

**Comment by elena-vi** [comment]
[Step 1 - UI send a Letter](https://trello.com/c/GvGYzdkE/153-step-1-ui-send-a-letter)

File: lib/hackney/pdf/preview.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+      end
+
+      private
+
+      def get_leasehold_info(payment_ref)
+        @leasehold_information_gateway.execute(payment_ref: payment_ref).first
+      end
+
+      def get_template_by_id(template_id)
+        @get_templates_gateway.execute.select { |temp| temp[:id] == template_id }.first
```
Comment (alexdesi) [inline lib/hackney/pdf/preview.rb]: I think there are a ruby method that stop when if find the first one (so you do not need to use `.first`, I might want to use that.

File: lib/hackney/pdf/preview_generator.rb [hunk 7]
```diff
@@ -1,27 +1,24 @@
...
-        correspondence_address_two = letter_params.fetch(:correspondence_address_two)
-        correspondence_address_three = letter_params.fetch(:correspondence_address_three)
+        correspondence_address_one = letter_params.fetch(:correspondence_address_1)
+        correspondence_address_two = letter_params.fetch(:correspondence_address_2)
+        correspondence_address_three = letter_params.fetch(:correspondence_address_3)
         correspondence_postcode = letter_params.fetch(:correspondence_postcode)
         lessee_short_name = letter_params.fetch(:lessee_short_name)
         property_address = letter_params.fetch(:property_address)
         arrears_letter_1_date = letter_params.fetch(:arrears_letter_1_date)
         total_collectable_arrears_balance = letter_params.fetch(:total_collectable_arrears_balance)
```
Comment (alexdesi) [inline lib/hackney/pdf/preview_generator.rb]: as discussed before all these variables passed into the template sound like a code smell.
Go for it but it should refactor those at some point.

File: lib/hackney/pdf/use_case_factory.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
+module Hackney
+  module PDF
+    class UseCaseFactory
+      TEMPLATE_DIRECTORY_PATH = 'lib/hackney/pdf/templates/'.freeze
```
Comment (alexdesi) [inline lib/hackney/pdf/use_case_factory.rb]: 🎁

File: spec/lib/hackney/pdf/get_templates_spec.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+
+describe Hackney::PDF::GetTemplates do
+  subject do
+    described_class.new(
+      template_directory_path: 'spec/lib/hackney/pdf/'
+    )
+  end
+
+  it 'the templates are found' do
```
Comment (alexdesi) [inline spec/lib/hackney/pdf/get_templates_spec.rb]: better to say  `gets the templates` or `retrieves...` or `finds...`

File: lib/hackney/pdf/preview_generator.rb (hunk 7 above)
Comment (elena-vi) [reply lib/hackney/pdf/preview_generator.rb]: sooooonnnn 😄

File: spec/lib/hackney/pdf/get_templates_spec.rb (hunk 9 above)
Comment (elena-vi) [reply spec/lib/hackney/pdf/get_templates_spec.rb]: yea muy bad

File: lib/hackney/service_charge/gateway/service_charge_gateway.rb [hunk 10]
```diff
@@ -39,11 +41,61 @@ def get_cases_by_refs(refs)
           end
         end
 
+        def fake_get_cases_by_refs(refs)
+          pp refs.first
+          raise Hackney::ServiceCharge::Exceptions::ServiceChargeException, refs if refs.first == '123'
```
Comment (alexdesi) [inline lib/hackney/service_charge/gateway/service_charge_gateway.rb]: as discussed, '123' looks like a magic number, please add a constant if you want to use it, or consider to change logic

File: lib/hackney/service_charge/gateway/service_charge_gateway.rb [hunk 11]
```diff
@@ -39,11 +41,61 @@ def get_cases_by_refs(refs)
           end
         end
 
+        def fake_get_cases_by_refs(refs)
+          pp refs.first
```
Comment (alexdesi) [inline lib/hackney/service_charge/gateway/service_charge_gateway.rb]: this should be removed.

**Pull request merged**